lib/spack/spack/binary_distribution.py

```diff
@@ -180,7 +181,11 @@ def write_buildinfo_file(spec, workdir, rel=False):
                         tty.warn(msg)
 
             if relocate.needs_binary_relocation(m_type, m_subtype):
-                if not filename.endswith('.o'):
+                if ((m_subtype in ('x-executable', 'x-sharedlib')
+                    and sys.platform != 'darwin') or
+                   (m_subtype in ('x-mach-binary')
+                    and sys.platform == 'darwin') or
+                   (not filename.endswith('.o'))):
```

Comment:
I'm not too familiar with the internals of relocation, but I did peek and see that `relocate.py` has a `file_is_relocatable()` method which looks at the type and subtype.  Should this logic go in there instead?  Otherwise, this looks reasonable to me.

Reply:
file_is_relocatable uses the same logic to determine how to get the rpaths and then discards them from the set up strings in the binary. It's not equivalent.
https://github.com/spack/spack/blob/03790a4f3609f1bedb7dee947c8712b0ab1e3348/lib/spack/spack/relocate.py#L927